app/services/media_storage/abstract_adapter.rb

```diff
@@ -12,14 +12,21 @@ def stored_path(content_type, media_type, *path_prefix)
       raise NotImplementedError
     end
 
+    # Returns URL/path to be used to make a GET request to the storage service
     def get_path(path, opts={})
       raise NotImplementedError
     end
 
+    # Returns URL/path to be used to make a PUT request to the storage service
     def put_path(path, opts={})
       raise NotImplementedError
     end
 
+    # Returns required headers for making a put request to the storage service
+    def headers_for_direct_upload
```

Comment:
is this abstract across all adapter implementations? If not it's probably a private? method on the azure adapter only

Reply:
If we provide content-type, content-disposition etc then this method would exist in each concrete implementation class, so this would be an abstract method. e.g https://github.com/rails/rails/blob/f16431e0c04a72e5a42fe1f7c5b5aefe76220ddc/activestorage/lib/active_storage/service.rb#L136-L138